fs/fcb/src/fcb.c

```diff
@@ -265,3 +301,183 @@ fcb_clear(struct fcb *fcb)
     }
     return rc;
 }
+
+int
+fcb_init_flash_area(struct fcb *fcb, int flash_area_id, uint8_t version)
+{
+    const struct flash_area *fa;
+    struct sector_range *sector_ranges;
+    int sector_range_cnt = 0;
+    int rc;
+
+    /*
+     * We don't know how big the area is so need to check how many sectors are
+     * there and then read information about all sectors - this is needed to
+     * properly initialize FCB. Critical log is somewhat important and shall be
+     * created, so we just assert on any error.
+     *
+     * XXX Should we do something else here?
+     */
+    rc = flash_area_to_sector_ranges(flash_area_id, &sector_range_cnt, NULL);
+    assert(rc == 0 && sector_range_cnt > 0);
+    sector_ranges = malloc(sizeof(struct sector_range) * sector_range_cnt);
+    assert(sector_ranges);
+    rc = flash_area_to_sector_ranges(flash_area_id, &sector_range_cnt, sector_ranges);
+    assert(rc == 0 && sector_range_cnt > 0);
+
+    fcb->f_ranges = sector_ranges;
+    fcb->f_range_cnt = sector_range_cnt;
+    fcb->f_sector_cnt = sector_ranges[sector_range_cnt - 1].sr_first_sector +
+        sector_ranges[sector_range_cnt - 1].sr_sector_count;
+    fcb->f_magic = 0xDEADBEEF;
+    fcb->f_version = version;
+
+    /*
+     * Initialize log in dedicated flash area. This has to succeed since it
+     * should be in dedicated flash area and nothing should prevent us from
+     * creating log there.
+     */
+    rc = fcb_init(fcb);
+    if (rc) {
+        /* Need to erase full area here */
+        rc = flash_area_open(flash_area_id, &fa);
+        assert(rc == 0);
+
+        flash_area_erase(fa, 0, fa->fa_size);
+        rc = fcb_init(fcb);
+        assert(rc == 0);
+    }
+
+    return rc;
+}
+
+int
+fcb_get_sector_info(const struct fcb *fcb, int sector,
+    struct fcb_sector_info *info)
+{
+    struct sector_range *srp = fcb->f_ranges;
+    int i;
+
+    if (sector == FCB_SECTOR_OLDEST) {
+        sector = fcb->f_oldest_sec;
+    }
+
+    for (i = 0; i < fcb->f_range_cnt; ++i, ++srp) {
+        if (srp->sr_sector_count <= sector) {
+            sector -= srp->sr_sector_count;
+            continue;
+        }
+        info->si_range = srp;
+        info->si_sector_in_range = sector;
+        info->si_sector_offset = srp->sr_range_start +
+            sector * srp->sr_sector_size;
+        return 0;
+    }
+    return FCB_ERR_ARGS;
+}
+
+int
+fcb_get_sector_loc(const struct fcb *fcb, int sector, struct fcb_entry *entry)
+{
+    struct fcb_sector_info info;
+    int rc = 0;
+
+    if (sector == FCB_SECTOR_OLDEST) {
+        sector = fcb->f_oldest_sec;
+    }
+    rc = fcb_get_sector_info(fcb, sector, &info);
+    if (rc == 0) {
+        entry->fe_range = info.si_range;
+        entry->fe_sector = (uint16_t)sector;
+    }
+
+    return rc;
+}
+
+
+int
+fcb_offset_in_sector(const struct fcb *fcb, int offset, int sector)
```

Comment:
This function (`fcb_offset_in_sector`) would probably be simpler if it called `fcb_get_sector_info()`.

Reply:
done